### ISSUE TITLE
reopen issue #143: 

### ISSUE THREAD

**Issue by kwtalley** [orika-mapper/orika#105]
Found this older issue which I am seeing in the latest release: **1.4.6**

[https://code.google.com/archive/p/orika/issues/143](url)

Test case from the post that shows the issue:

```
package ma.glasnost.orika.test.community;

import ma.glasnost.orika.MapperFacade;
import ma.glasnost.orika.MapperFactory;
import ma.glasnost.orika.impl.DefaultMapperFactory;
import org.junit.Assert;
import org.junit.Test;

public class AbstractCircularReferenceTestCase {
    @Test
    public void test() {
        DefaultMapperFactory.Builder builder = new DefaultMapperFactory.Builder();
        MapperFactory factory = builder.build();
        factory.classMap(Entity.class, BaseEntity.class).field("anotherEntity", "anotherBaseEntity").register();
        factory.classMap(AnotherEntity.class, AnotherBaseEntity.class).field("abstractEntity", "abstractBaseEntity").register();

        MapperFacade mapperFacade = factory.getMapperFacade();

        Entity entity = new Entity();

        AnotherEntity anotherEntity = new AnotherEntity();
        anotherEntity.setAbstractEntity(entity);

        entity.setAnotherEntity(anotherEntity);


        BaseEntity baseEntity = mapperFacade.map(entity, BaseEntity.class);

        Assert.assertEquals(entity, entity.getAnotherEntity().getAbstractEntity());
        Assert.assertEquals(baseEntity, baseEntity.getAnotherBaseEntity().getAbstractBaseEntity());
    }

    public static abstract class AbstractEntity {
        AnotherEntity anotherEntity;

        public AnotherEntity getAnotherEntity() {
            return anotherEntity;
        }

        public void setAnotherEntity(AnotherEntity anotherEntity) {
            this.anotherEntity = anotherEntity;
        }
    }

    public static class Entity extends AbstractEntity {
    }

    public static class AnotherEntity {
        AbstractEntity abstractEntity;

        public AbstractEntity getAbstractEntity() {
            return abstractEntity;
        }

        public void setAbstractEntity(AbstractEntity abstractEntity) {
            this.abstractEntity = abstractEntity;
        }
    }

    public static class AbstractBaseEntity {
        AnotherBaseEntity anotherBaseEntity;

        public AnotherBaseEntity getAnotherBaseEntity() {
            return anotherBaseEntity;
        }

        public void setAnotherBaseEntity(AnotherBaseEntity anotherBaseEntity) {
            this.anotherBaseEntity = anotherBaseEntity;
        }
    }

    public static class BaseEntity extends AbstractBaseEntity {
    }

    public static class AnotherBaseEntity {
        AbstractBaseEntity abstractBaseEntity;

        public AbstractBaseEntity getAbstractBaseEntity() {
            return abstractBaseEntity;
        }

        public void setAbstractBaseEntity(AbstractBaseEntity abstractBaseEntity) {
            this.abstractBaseEntity = abstractBaseEntity;
        }
    }
}
```

I can see that converting a list of abstract types will:
1. check the typeCache keying off abstract type
2. fail to find it and resolve the concrete destination type
3. check the typeCache again with correct destination type and is successful

however single objects (use aToB converting) do not attempt to resolve the concrete destination type until after checking the typeCache:
1. check the typeCache keying off abstract type
2. fail to find it and create a new instance (basically a copy)
3. put new instance in the typeCache (now typeCache has 2 objects that are identical in contents but different instances)

**Comment by andreabertagnolli** [comment]
I tried the test case, it breaks because it can't verify equality between baseEntity and abstractBaseEntity into another baseEntity, but it's because they are different objects!
so i tried to override an equal method on AbstractBaseEntity and it works.

maybe I didn't understand...

**Comment by andreabertagnolli** [comment]
I tried with current head (what will be version 1.5.0)

**Comment by brabenetz** [comment]
I think equality is not enough.
The Problem is, that the two Objects should be exactly the same instance.

A small update in UseCustomMapperStrategy solves the problem:
```
diff --git core/src/main/java/ma/glasnost/orika/impl/mapping/strategy/UseCustomMapperStrategy.java core/src/main/java/ma/glasnost/orika/impl/mapping/strategy/UseCustomMapperStrategy.java
index e3a95d7..98b5fc9 100644
--- core/src/main/java/ma/glasnost/orika/impl/mapping/strategy/UseCustomMapperStrategy.java
+++ core/src/main/java/ma/glasnost/orika/impl/mapping/strategy/UseCustomMapperStrategy.java
@@ -58,6 +58,11 @@ public abstract class UseCustomMapperStrategy extends AbstractMappingStrategy {
         
         Object resolvedSourceObject = unenhancer.unenhanceObject(sourceObject, sourceType);
         
+        Object cachedObject = context.getMappedObject(sourceObject, destinationType);
+        if (cachedObject != null) {
+            return cachedObject;
+        }
+        
         Object newInstance = getInstance(resolvedSourceObject, destinationObject, context);
         
         context.cacheMappedObject(sourceObject, destinationType, newInstance);

```

I also found some legacy code which (I'm pretty sure) can be removed:
In MappingContext.java is a code snippets with the comment "// Quick fix for Issue 68".
The Issue68TestCase.java works fine without it.

```
  // Quick fix for Issue 68
  for (Type<Object> t : (Type<Object>[]) destinationType.getInterfaces()) {
    cacheMappedObject(source, t, destination);
  }
```

**Comment by andreabertagnolli** [comment]
Will be fixed on 1.5.0
